File: frontend/src/schemas/carousel.ts
import { z } from "zod";

export const carouselDesignColorsSchema = z.object({
  primary: z.string(),
  accent: z.string(),
  bg: z.string(),
  text: z.string(),
  text_muted: z.string(),
  text_dim: z.string(),
  border: z.string(),
  glow: z.string(),
});

export const carouselDesignTypographySchema = z.object({
  font_family_heading: z.string(),
  font_family_body: z.string(),
  font_family_badge: z.string(),
});

export const carouselBlogImageMapEntrySchema = z.object({
  slide_number: z.number(),
  heading: z.string(),
  alt: z.string(),
});

export const carouselDesignImagesSchema = z.object({
  hero: z.string(),
  slides: z.array(z.string()),
  rendered_slides_pt: z.array(z.string()).nullable().optional(),
  rendered_slides_en: z.array(z.string()).nullable().optional(),
  blog_image_map: z
    .array(carouselBlogImageMapEntrySchema)
    .nullable()
    .optional(),
});

export const carouselDesignLayoutSchema = z.object({
  badge_label: z.string(),
  swipe_text: z.string(),
  progress_segments: z.number(),
});

export const carouselDesignResponseSchema = z.object({
  colors: carouselDesignColorsSchema,
  typography: carouselDesignTypographySchema,
  images: carouselDesignImagesSchema,
  layout: carouselDesignLayoutSchema,
  theme_name: z.string(),
});

export const carouselBlogI18nResponseSchema = z.object({
  markdown: z.string(),
  title: z.string(),
  subtitle: z.string().nullable(),
  language: z.string(),
  available_languages: z.array(z.string()),
});

// Same shape as the i18n response plus a design block (AE-0154 dedup).
export const carouselBlogWithDesignResponseSchema =
  carouselBlogI18nResponseSchema.extend({
    design: carouselDesignResponseSchema,
  });

export const carouselProjectResponseSchema = z.object({
  id: z.string(),
  topic: z.string(),
  audience: z.string(),
  niche: z.string(),
  title: z.string().nullable(),
  subtitle: z.string().nullable(),
  title_en: z.string().nullable().optional(),
  subtitle_en: z.string().nullable().optional(),
  theme: z.string(),
  status: z.string(),
  image_model: z.string().optional(),
  image_style: z.string().optional(),
  primary_color: z.string().nullable(),
  accent_color: z.string().nullable(),
  background_color: z.string().nullable(),
  is_public: z.boolean().optional(),
  current_phase: z.string().nullable().optional(),
  phase_status: z.string().nullable().optional(),
  error_message: z.string().nullable().optional(),
  output_dir: z.string().nullable().optional(),
  research_sources: z.array(z.unknown()).optional(),
  slides: z.array(z.unknown()).optional(),
  blog_markdown: z.string().nullable(),
  blog_translations: z.record(z.string(), z.string()).nullable().optional(),
  caption: z.string().nullable(),
  linkedin_post_pt: z.string().nullable().optional(),
  linkedin_post_en: z.string().nullable().optional(),
  pdf_path: z.string().nullable().optional(),
  pdf_path_en: z.string().nullable().optional(),
  design_tokens: z.unknown().nullable().optional(),
  creator_name: z.string().nullable().optional(),
  creator_handle: z.string().nullable().optional(),
  creator_avatar_url: z.string().nullable().optional(),
  template_version: z.string().nullable().optional(),
  slide_layout_strategy: z.string().nullable().optional(),
  created_at: z.string(),
  updated_at: z.string(),
});

export const carouselProjectListResponseSchema = z.object({
  items: z.array(carouselProjectResponseSchema),
  total: z.number(),
  limit: z.number(),
  offset: z.number(),
});

export const carouselSlideResponseSchema = z.object({
  id: z.string(),
  slide_number: z.number(),
  slide_type: z.string(),
  heading: z.string(),
  body: z.string(),
  image_path: z.string().nullable().optional(),
  created_at: z.string(),
  updated_at: z.string(),
});

/**
 * Only these (model, style) tuples are wired in the backend registry.
 * Anything else is rejected by the API with 422 before the pipeline runs.
 */
export const SUPPORTED_IMAGE_COMBOS: ReadonlyArray<readonly [string, string]> =
  [
    ["gemini", "comic_neon"],
    ["openai", "cinematic"],
    ["openai", "hyperreal"],
    ["openai", "neo_anime"],
    ["openai", "flat_editorial"],
  ];

export const carouselCreateRequestSchema = z
  .object({
    topic: z.string().min(1).max(500),
    audience: z.string().min(1).max(500),
    niche: z.string().min(1).max(200),
    // root key | "auto" | custom-palette UUID (36 chars). Matches the backend
    // varchar(64) widened in AE-0268 so custom-palette references validate.
    theme: z.string().max(64).default("auto"),
    image_model: z.enum(["gemini", "openai"]).default("openai"),
    image_style: z
      .enum([
        "comic_neon",
        "cinematic",
        "hyperreal",
        "neo_anime",
        "flat_editorial",
      ])
      .default("neo_anime"),
    strategy: z.string().max(50).optional(),
  })
  .refine(
    (data) =>
      SUPPORTED_IMAGE_COMBOS.some(
        ([m, s]) => m === data.image_model && s === data.image_style,
      ),
    {
      message: "image_model + image_style combination is not supported",
      path: ["image_style"],
    },
  );

export const carouselPhaseProgressSlideSchema = z.object({
  number: z.number(),
  status: z.enum(["pending", "in_flight", "done", "failed"]),
  style: z.string().optional(),
  scene: z.string().optional(),
});

export const carouselPhaseProgressSchema = z.object({
  phase: z.string(),
  label: z.string(),
  current: z.number().optional(),
  total: z.number().optional(),
  detail: z.string().optional(),
  slides: z.array(carouselPhaseProgressSlideSchema).optional(),
});

export type CarouselPhaseProgress = z.infer<typeof carouselPhaseProgressSchema>;

export type CarouselBlogImageMapEntry = z.infer<
  typeof carouselBlogImageMapEntrySchema
>;
export type CarouselDesignResponse = z.infer<
  typeof carouselDesignResponseSchema
>;
export type CarouselBlogI18nResponse = z.infer<
  typeof carouselBlogI18nResponseSchema
>;
export type CarouselBlogWithDesignResponse = z.infer<
  typeof carouselBlogWithDesignResponseSchema
>;
export type CarouselProjectResponse = z.infer<
  typeof carouselProjectResponseSchema
>;
export type CarouselProjectListResponse = z.infer<
  typeof carouselProjectListResponseSchema
>;
export type CarouselSlideResponse = z.infer<typeof carouselSlideResponseSchema>;
export type CarouselCreateRequest = z.infer<typeof carouselCreateRequestSchema>;

File: backend/src/rag_backend/api/schemas/carousel.py
"""Carousel Pydantic schemas."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field, field_validator, model_validator

from rag_backend.domain.constants import (
    IMAGE_MODEL_DEFAULT,
    IMAGE_STYLE_DEFAULT,
    SUPPORTED_IMAGE_COMBOS,
    VALID_IMAGE_MODELS,
    VALID_IMAGE_STYLES,
)

_ERR_INVALID_IMAGE_MODEL = "image_model must be one of {}, got {!r}"
_ERR_INVALID_IMAGE_STYLE = "image_style must be one of {}, got {!r}"
_ERR_UNSUPPORTED_IMAGE_COMBO = (
    "image_model={!r} with image_style={!r} is not supported. Allowed: {}"
)


class CarouselProjectCreate(BaseModel):
    topic: str = Field(..., min_length=1, max_length=500)
    audience: str = Field(..., min_length=1, max_length=500)
    niche: str = Field(..., min_length=1, max_length=200)
    slides_config: str = Field(default="1 intro, 3 content, 1 closing, 1 cta", max_length=200)
    language: str = Field(default="pt-BR", max_length=10)
    generate_images: bool = True
    theme: str = Field(default="auto", max_length=30)
    image_model: str = Field(default=IMAGE_MODEL_DEFAULT, max_length=30)
    image_style: str = Field(default=IMAGE_STYLE_DEFAULT, max_length=30)

    @field_validator("image_model")
    @classmethod
    def _check_image_model(cls, value: str) -> str:
        if value not in VALID_IMAGE_MODELS:
            raise ValueError(_ERR_INVALID_IMAGE_MODEL.format(sorted(VALID_IMAGE_MODELS), value))
        return value

    @field_validator("image_style")
    @classmethod
    def _check_image_style(cls, value: str) -> str:
        if value not in VALID_IMAGE_STYLES:
            raise ValueError(_ERR_INVALID_IMAGE_STYLE.format(sorted(VALID_IMAGE_STYLES), value))
        return value

    @model_validator(mode="after")
    def _check_combo(self) -> "CarouselProjectCreate":
        if (self.image_model, self.image_style) not in SUPPORTED_IMAGE_COMBOS:
            raise ValueError(
                _ERR_UNSUPPORTED_IMAGE_COMBO.format(
                    self.image_model, self.image_style, sorted(SUPPORTED_IMAGE_COMBOS)
                )
            )
        return self


class CarouselProjectUpdate(BaseModel):
    title: str | None = Field(None, max_length=500)
    subtitle: str | None = None
    blog_markdown: str | None = None
    caption: str | None = None


class CarouselSlideResponse(BaseModel):
    id: UUID
    slide_number: int
    slide_type: str
    heading: str
    body: str
    image_path: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ResearchSourceResponse(BaseModel):
    id: UUID
    source_url: str
    source_type: str
    title: str | None = None
    relevance_score: float
    created_at: datetime

    model_config = {"from_attributes": True}


class CarouselProjectResponse(BaseModel):
    id: UUID
    topic: str
    audience: str
    niche: str
    title: str | None
    subtitle: str | None
    title_en: str | None = None
    subtitle_en: str | None = None
    theme: str
    image_model: str = IMAGE_MODEL_DEFAULT
    image_style: str = IMAGE_STYLE_DEFAULT
    primary_color: str | None
    accent_color: str | None
    background_color: str | None
    blog_markdown: str | None
    blog_translations: dict[str, str] | None = None
    caption: str | None
    linkedin_post_pt: str | None = None
    linkedin_post_en: str | None = None
    design_tokens: dict[str, dict[str, str | int | list[str] | None]] | None = None
    status: str
    error_message: str | None = None
    output_dir: str | None = None
    pdf_path: str | None = None
    pdf_path_en: str | None = None
    slides: list[CarouselSlideResponse] = Field(default_factory=list)
    research_sources: list[ResearchSourceResponse] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class CarouselProjectListResponse(BaseModel):
    items: list[CarouselProjectResponse]
    total: int
    limit: int
    offset: int


class CarouselStatusResponse(BaseModel):
    id: UUID
    status: str
    error_message: str | None = None
    phase_progress: dict[str, str | int | list[dict[str, str | int]]] | None = None
    updated_at: datetime

    model_config = {"from_attributes": True}


class CarouselGenerateRequest(BaseModel):
    sources: list[str] | None = Field(default=None, description="Optional source URLs to research")


class InstagramPublishRequest(BaseModel):
    caption: str = Field(..., min_length=1, max_length=2200)


class InstagramPublishResponse(BaseModel):
    status: str
    ig_post_id: str | None = None
    error_message: str | None = None


class CarouselCaptionResponse(BaseModel):
    caption: str
    hashtags: list[str]


class CarouselBlogResponse(BaseModel):
    markdown: str
    title: str
    subtitle: str | None = None


class CarouselBlogI18nResponse(BaseModel):
    markdown: str
    title: str
    subtitle: str | None
    language: str
    available_languages: list[str]

    model_config = {"from_attributes": False}


class CarouselDesignColors(BaseModel):
    primary: str
    accent: str
    bg: str
    text: str
    text_muted: str
    text_dim: str
    border: str
    glow: str


class CarouselDesignTypography(BaseModel):
    font_family_heading: str
    font_family_body: str
    font_family_badge: str


class CarouselBlogImageMapEntry(BaseModel):
    slide_number: int
    heading: str
    alt: str


class CarouselDesignImages(BaseModel):
    hero: str
    slides: list[str]
    rendered_slides_pt: list[str] | None = None
    rendered_slides_en: list[str] | None = None
    blog_image_map: list[CarouselBlogImageMapEntry] | None = None


class CarouselDesignLayout(BaseModel):
    badge_label: str
    swipe_text: str
    progress_segments: int


class CarouselDesignResponse(BaseModel):
    colors: CarouselDesignColors
    typography: CarouselDesignTypography
    images: CarouselDesignImages
    layout: CarouselDesignLayout
    theme_name: str

    model_config = {"from_attributes": False}


class CarouselBlogWithDesignResponse(BaseModel):
    markdown: str
    title: str
    subtitle: str | None
    language: str
    available_languages: list[str]
    design: CarouselDesignResponse

    model_config = {"from_attributes": False}
